### ISSUE TITLE
Remove "Campaigns" section in favor of Relationships

### ISSUE THREAD

**Issue by mgoffin** [crits/crits#446]
Now that Relationships have a "confidence" and "reason" there is basically no reason to keep the "Campaigns" section on the Details page around as a feature. We should use Relationships to relate a TLO to a Campaign.

One issue with this is that Relationships are a bit cumbersome to add. One would have to go to a TLO, copy the ID to the clipboard, then go to the Campaign and add the relationship.

What I  propose is that we do the following:

- Remove the "campaigns" TLO attribute.
  - Which means we can remove the Embedded Campaign document stuff as well as one of the classes that gets inherited for adding campaign attribution to a TLO.
- Migrate existing Attribution to Relationships using migrate.
  - Also add this to the "upgrade" management command to make this easy for people to do.
- Remove the "Campaigns" section of the Details page.
- Remove all of the tabs on the Campaign Details page for the different TLOs.
- Add a button to the top-right (near Favorites) which is "Attribute Campaign".
  - Or make it another button next to the "Add Relationship" button?
  - The button would bring up a form similar to the current "Add Campaign" button.
    - We would ask for a:
      - Campaign (populated dropdown list) - Required
      - Relationship Type (populated dropdown list, with "Related To" as the default) - Required
      - Confidence (populated dropdown based on the relationship confidence values) - Required
      - Reason (open text field) - Optional
  - The form would submit but would now add a Relationship and updated the "Relationships" section of the Details page automatically.
- Remove the Campaign mapreduce since it will no longer be necessary.
- Update the Campaign listing page so that it shows counts of each TLO that it is related to.
- Update the listing pages of other TLOs which currently show the list of Campaigns so that it gets this information from Relationships.
- Update the "Relationships" service to remove the special need to check the "campaigns" attribute, and let it run completely off of Relationships.

This would also wind up fixing #443.

Anything else I might have missed trying to make this work?

**Comment by thelok** [comment]
So... my investigation into relationships back a while ago.

The code on the back end for loading relationships is not efficient. Let's say you have 1000 domains attributed to a campaign. Fine, each domain will have a single relationship to the campaign. But open up the details page for the campaign and it will have 1000 relationships -- it should load take a long time to process. It does some sorting and I forget where that code was located...

As long as testing is done with scale in mind and the UI is still easy to use then I don't have complaints.

**Comment by mgoffin** [comment]
If anyone has any ideas for how to make it more efficient i'm all about it :)

**Comment by brlogan** [comment]
It has always bothered me that you could seemingly relate a TLO to a Campaign in two different ways, so from that perspective I like this a lot.

However, this change makes me a little nervous because Campaigns typically have a lot more related to them than other TLOs. The current tabbed view with the ability to sort and filter can be pretty useful for large data sets, and they load much more quickly. But with the standard relationship view, we lose functionality, end up with a very long page, and the page takes noticeably longer to load.
Could we switch to using relationships for Campaigns, but keep the tabbed with jtables view for the Campaign TLO details page?

**Comment by wxsBSD** [comment]
I think the campaigns listing staying as table with counts per item is something Mike wants to keep, for the reasons you cite. I know I would be unhappy if that changed.

**Comment by brlogan** [comment]
Oh good!
This was in his comment: "Remove all of the tabs on the Campaign Details page for the different TLOs."
But maybe I misunderstood.

**Comment by wxsBSD** [comment]
Oh, that would certainly suck. You do bring up a good point, so thank you!

**Comment by mgoffin** [comment]
Wes is misunderstanding you. My proposal is to remove all the tabs on the Details page, but keep all the columns on the listing page.

We could, if it would make life easier, have a single tab for a "jtable view" of the Related TLOs. You can pick which TLO type and it can render the table as they look and act now?

**Comment by wxsBSD** [comment]
I like the listing page as it is is great, and having a table where you can select the TLO and have it render is also good.

**Comment by brlogan** [comment]
Ah, I see where the miscommunication is. Sorry for the confusion @wxsBSD. I was talking about the Campaigns **Details** page, not the Campaigns **List** page. I also like the list page how it is.

A "Table View" tab would address the issue of being able to sort and filter the attributed TLOs; I think that's a good compromise.  Hopefully we can figure out some optimizations because I do foresee slow page load being an issue. It's already a small issue for some of our Events that have many relationships.

**Comment by dsnellgrove** [comment]
TL;DR: without significant changes in relationship handling this change is incompatible with my existing CRITs instance

Having run into objects with tens of thousands of relationships in them the page load is slow on the server side (mongoengine loading a multi-megabyte document like that is quite slow; building the django template is also slow and quite memory intensive), the resulting tens of megabytes of html sent to the browser is also slow (rendering this results in a page that locks up for over a minute on initial load and with almost any interaction made with it). The proposed change would result in more than an order of magnitude greater number of relationships centering on a single object compared to this. You would need to make relationships and the details page rendering (server and client side) about two orders of magnitude more performant to get anywhere near usable at my dataset's scale.

The other issue is the document size limit. Given relationships are bi-directional and the  current relationship embedded document size the outer limits of the number of objects you can attribute to a given campaign (this is a really outer limit that assumes they don't ever fill in a reason and have short usernames etc) is ~100,000 before you reach the 16MB barrier. That again doesn't come close to fitting my existing dataset in it.

**Comment by mgoffin** [comment]
@dsnellgrove All very valid points. I think the issue you (and others) are having will definitely be exacerbated by requiring even more Relationships (even though you can already use Relationships for Campaigns).

I suspect our "first-step solution" to this separate (I am considering it separate since it already exists) problem is going to lie in the template rendering situation. The relationship templates are especially nasty and bloated. There's lots of conditionals and loops all over the place and they do generate a fair bit of HTML to return. It might require us to take a more AJAX-y approach where we only fetch and render content as it is requested to be viewed. I am a huge fan of "having information immediately" like we do now so I can quickly glance at all of the details. Maybe a common-ground will be to only render the first N number of each TLO type and provide a "More..." link to get the next group?

**Comment by dsnellgrove** [comment]
Large objects can get pretty hairy just in mongoengine itself. Calling .save() on a mongoengine document with ~50,000 relationships took 30 seconds of CPU time in python last time I tried it. That's just mongoengine processing the .save() in python, not waiting on mongodb, not django/template time. This has to be repeated every single time to add, remove, edit relationships or otherwise change the campaign/object. In my instance someone was relating a ton of Indicators to another TLO and asked me why the job had been running for 12 hours and still wasn't done. After some profiling it turned out the mongoengine library itself was the bottleneck :(

Changes in the way the data is stored would also have to happen to allow scaling to the level using campaigns as a relationship node like this would require.

**Comment by mgoffin** [comment]
Yea. I was thinking about how much it would suck (or not) to move Relationships to their own collection. Maybe a conversation we need to have before we attempt to move forward with this Issue?

**Comment by TalonStryk** [comment]
After reading the above comments, I have a few concerns based on our work practices.  We are new to CRITS, so we may be doing things "the wrong way".

We are using Campaign TLO's to represent the intel reports/watchlists instead of "events" we receive for the following reasons:

1. Our analysts like the tabs across the top of the Campaign Details page with the various counts;  I've been asked if it was possible to do something similar on the Events detail page.  I understand that this is preference more than anything, and the Related TLO's tab with selectable types would work as well.  I'd also suggest having the counts from the Campaign listings page represented somewhere on the campaign details page to get an "at a glance" review of the campaign.
2.  When using the Restful API to integrate with other applications, we are currently utilizing indicators' , domains', IPs', and samples' Campaign attributes to supply the watch list, report, etc that the item came from.  There may be a better way to do this with the Restful API as it exists today.
3.  Having the Campaign attribute is useful when searching or looking at the various TLO listing pages to quickly identify what Campaigns a TLO is related to is very useful.  If a similar field on the listing jtables could be created using the relationship data, that would work as well.
4.  It is not currently possible to related TLO's such as samples, domains, indicators, and IPs to an event using the bulk upload options in the GUI.  If the TLO campaign attribute was removed, the same issue would occur with relating bulk uploaded TLO's to campaigns.  Adding the ability to create relationships in bulk, particularly on the bulk upload forms would resolve this issue.

**Comment by wxsBSD** [comment]
@mgoffin: one thing we may be able to do to make your idea of relationships being collected on demand is have a miniature embedded table interface that would have tabs for each elated type. I'm also wondering if we can do tricks to make collected objects faster in pymongo in this case.

**Comment by ckane** [comment]
I came across this thread earlier, and am now looking at the code in crits_mongoengine.py that implements get_relationship_objects - which I suspect to be the code that is called to render the per-related-TLO tables in different TLOs (so that summary fields can be displayed for related items in a foreign TLO).

I notice that it iterates linearly across the relationships, (I think) executing a new mongo query for each related TLO in the list. I suspect a more efficient approach might be to identify the distinct list of the "types" among the relationships, and run a query incorporating {'_id': {'$in': [... ObjectId(...), ObjectId(...) ...]}} for each of the related data types, rather than for each item in the relationship list.

This might speed up the page load time mentioned in comments by @dsnellgrove above, where there are many related items.

**Comment by ckane** [comment]
Check out PR #930 and lmk what you think
